core/types/celo_transaction_signing.go

```diff
@@ -17,94 +17,114 @@
-	x, ok := s2.(cel2Signer)
-	return ok && x.chainId.Cmp(s.chainId) == 0
+// latestCeloSigner creates a new celoSigner that is configured to handle all
+// celo forks for non celo transaction types it will delegate to the given
+// upstream signer.
+func latestCeloSigner(chainID *big.Int, upstreamSigner Signer) Signer {
+	return &celoSigner{
+		chainID:        chainID,
+		upstreamSigner: upstreamSigner,
+		activatedForks: forks,
+	}
 }
 
-func (s cel2Signer) SignatureValues(tx *Transaction, sig []byte) (R, S, V *big.Int, err error) {
-	if tx.Type() != CeloDynamicFeeTxV2Type && tx.Type() != CeloDenominatedTxType {
-		return s.londonSigner.SignatureValues(tx, sig)
+// Sender implements Signer.
+func (c *celoSigner) Sender(tx *Transaction) (common.Address, error) {
+	if funcs := c.findTxFuncs(tx); funcs != nil {
+		return funcs.sender(tx, funcs.hash, c.ChainID())
 	}
+	return c.upstreamSigner.Sender(tx)
+}
 
-	// Check that chain ID of tx matches the signer. We also accept ID zero here,
-	// because it indicates that the chain ID was not specified in the tx.
-	chainID := tx.inner.chainID()
-	if chainID.Sign() != 0 && chainID.Cmp(s.chainId) != 0 {
-		return nil, nil, nil, ErrInvalidChainId
+// SignatureValues implements Signer.
+func (c *celoSigner) SignatureValues(tx *Transaction, sig []byte) (r *big.Int, s *big.Int, v *big.Int, err error) {
+	if funcs := c.findTxFuncs(tx); funcs != nil {
+		return funcs.signatureValues(tx, sig, c.ChainID())
 	}
-	R, S, _ = decodeSignature(sig)
-	V = big.NewInt(int64(sig[64]))
-	return R, S, V, nil
+	return c.upstreamSigner.SignatureValues(tx, sig)
 }
 
-// Hash returns the hash to be signed by the sender.
-// It does not uniquely identify the transaction.
-func (s cel2Signer) Hash(tx *Transaction) common.Hash {
-	if tx.Type() == CeloDynamicFeeTxV2Type {
-		return prefixedRlpHash(
-			tx.Type(),
-			[]interface{}{
-				s.chainId,
-				tx.Nonce(),
-				tx.GasTipCap(),
-				tx.GasFeeCap(),
-				tx.Gas(),
-				tx.To(),
-				tx.Value(),
-				tx.Data(),
-				tx.AccessList(),
-				tx.FeeCurrency(),
-			})
+// Hash implements Signer.
+func (c *celoSigner) Hash(tx *Transaction) common.Hash {
+	if funcs := c.findTxFuncs(tx); funcs != nil {
+		return funcs.hash(tx, c.ChainID())
 	}
-	if tx.Type() == CeloDenominatedTxType {
-		return prefixedRlpHash(
-			tx.Type(),
-			[]interface{}{
-				s.chainId,
-				tx.Nonce(),
-				tx.GasTipCap(),
-				tx.GasFeeCap(),
-				tx.Gas(),
-				tx.To(),
-				tx.Value(),
-				tx.Data(),
-				tx.AccessList(),
-				tx.FeeCurrency(),
-				tx.MaxFeeInFeeCurrency(),
-			})
+	return c.upstreamSigner.Hash(tx)
+}
+
+// findTxFuncs returns the txFuncs for the given tx if it is supported by one
+// of the active forks. Note that this mechanism can be used to deprecate
+// support for tx types by having forks reutrn deprecatedTxFuncs for a tx type.
+func (c *celoSigner) findTxFuncs(tx *Transaction) *txFuncs {
+	// iterate in reverse over the activeForks and if any of them have a non nil txFuncs
```

Comment:
Why is reverse order important?

Reply:
We iterate in reverse order so we check the most recent forks first. This ensures that if a tx type has been deprecated we return the `deprecatedTxFuncs` rather than some valid funcs.

Reply:
Makes sense, then please add a comment on `forks` that it must be ordered.

Reply:
Updated and commented, I actually added a little functionality to forks to make things clearer.